cartofacade/queue.go

```diff
@@ -0,0 +1,186 @@
+	CartoType
+	// Nil represents the go type nil.
+	Nil
+	// Error represents the go type error.
+	Error
+)
+
+// Response defines the result of one piece of work that can be put on the result channel.
+type Response struct {
+	Result     interface{}
+	ResultType ResultType
+}
+
+// Queue represents a queue to consume work from and enforce one call into C at a time.
+type Queue struct {
+	WorkChannel     chan WorkItem
+	CartoLib        *cgoApi.CartoLibInterface
+	Carto           cgoApi.CartoInterface
+	CartoConfig     cgoApi.CartoConfig
+	CartoAlgoConfig cgoApi.CartoAlgoConfig
+}
+
+// WorkItemInterface defines one piece of work that can be put on the queue.
+type WorkItemInterface interface {
+	DoWork(q *Queue) (interface{}, ResultType, error)
+}
+
+// WorkItem defines one piece of work that can be put on the queue.
+type WorkItem struct {
+	Result   chan Response
+	workType WorkType
+	inputs   map[InputType]interface{}
+}
+
+// NewQueue instantiates the Queue struct.
+func NewQueue(cartoLib cgoApi.CartoLibInterface, cartoCfg cgoApi.CartoConfig, cartoAlgoCfg cgoApi.CartoAlgoConfig) Queue {
+	return Queue{
+		WorkChannel:     make(chan WorkItem),
+		Carto:           &cgoApi.Carto{},
+		CartoLib:        &cartoLib,
+		CartoConfig:     cartoCfg,
+		CartoAlgoConfig: cartoAlgoCfg,
+	}
+}
+
+// DoWork provides the logic to call the correct cgo functions with the correct input.
+func (w *WorkItem) DoWork(
+	q *Queue,
+) (interface{}, ResultType, error) {
+	switch w.workType {
+	case Initialize:
+		carto, err := cgoApi.New(q.CartoConfig, q.CartoAlgoConfig, *q.CartoLib)
+		return carto, CartoType, err
+	case Start:
+		return nil, Nil, q.Carto.Start()
+	case Stop:
+		return nil, Nil, q.Carto.Stop()
+	case Terminate:
+		return nil, Nil, q.Carto.Terminate()
+	case AddSensorReading:
+		reading, ok := w.inputs[Reading].([]byte)
+		if !ok {
+			return nil, Nil, errors.New("could not cast inputted byte to byte slice")
+		}
+
+		timestamp, ok := w.inputs[Timestamp].(time.Time)
+		if !ok {
+			return nil, Nil, errors.New("could not cast inputted timestamp to times.Time")
+		}
+
+		return nil, Nil, q.Carto.AddSensorReading(reading, timestamp)
+	case Position:
+		positionResponse, err := q.Carto.GetPosition()
+		return positionResponse, PositionResponse, err
+	case InternalState:
+		internalState, err := q.Carto.GetInternalState()
+		return internalState, ByteSlice, err
+	case PointCloudMap:
+		pointCloudMap, err := q.Carto.GetPointCloudMap()
+		return pointCloudMap, ByteSlice, err
+	}
+	return nil, Nil, fmt.Errorf("no worktype found for: %v", w.workType)
+}
+
+// Request puts incoming requests on the queue and consumes from queue.
+func (q *Queue) Request(ctxParent context.Context, workType WorkType, inputs map[InputType]interface{}, timeout time.Duration) Response {
+	ctx, cancel := context.WithTimeout(ctxParent, timeout)
+	defer cancel()
+
+	work := WorkItem{
+		Result:   make(chan Response, 1),
```

Comment:
@nicksanford A bunch of answers [here](https://stackoverflow.com/questions/8593645/is-it-ok-to-leave-a-channel-open) say that once there are no more references to the channel, it will be garbage collected. So I think we are safe here